llvm/test/CodeGen/WebAssembly/fpclamptosat.ll

```diff
@@ -524,9 +524,7 @@ define i64 @utest_f64i64(double %x) {
 ; CHECK-NEXT:    local.get 0
 ; CHECK-NEXT:    call __fixunsdfti
 ; CHECK-NEXT:    local.get 1
-; CHECK-NEXT:    i32.const 8
-; CHECK-NEXT:    i32.add
-; CHECK-NEXT:    i64.load 0
+; CHECK-NEXT:    i64.load 8
```

Comment:
I'm curious why these layout changes lead to extra folding here, since the alignment of i128 previously matched the alignment for i64 and these are i64 loads... actually I also don't know why the alignment should affect folding anyway. Are we missing a folding opportunity somewhere else too?

Reply:
hmm. in general, more folds are valid with naturally-aligned types, because if you see two random 4-byte-aligned types, using an 8-byte-aligned load is only valid if they're on the same boundary. LLVM tends to allow these sorts of things that are only relevant for some machines to creep into its global reasoning about optimization, and not with no reason: it's always safer to do the more conservative optimization that is valid on all machines.

so wasm's model might suggest it shouldn't be affected by this, but in practice LLVM reasons substantially about this. so I am not surprised a machine model that _should_ care less still optimizes better.

Reply:
in other words I'm betting something way upstream is bailing over-cautiously

Reply:
The issue is a missing nuw. Wasm load/store instructions' offsets don't perform wrapping, so constant folding adds into them usually requires the adds to have nuw or similar. In these tests, there's a 16-byte load that's getting split into two 8-byte loads, and the second one has a "+ 8" on its base address, so the question is, can we prove that add doesn't overflow. When the base alignment is known to be 16 bytes, then we can. But also, at a higher level, the legalization code should know that the original 16-byte load doesn't half wrap around, so it can just put nuw on the add up front, in which case we can always fold it.

I've now submittted https://github.com/llvm/llvm-project/pull/119288 to implement that.